File: aidbox-features/aidbox-orchestration-service/src/fhir-types/profile-helpers.ts
/**
 * Runtime helpers for generated FHIR profile classes.
 *
 * This file is copied verbatim into every generated TypeScript output and
 * imported by profile modules.  It provides:
 *
 * - **Slice helpers** – match, get, set, and default-fill array slices
 *   defined by a FHIR StructureDefinition.
 * - **Extension helpers** – read complex (nested) FHIR extensions into
 *   plain objects.
 * - **Choice-type helpers** – wrap/unwrap polymorphic `value[x]` fields so
 *   profile classes can expose a flat API.
 * - **Validation helpers** – lightweight structural checks that profile
 *   classes call from their `validate()` method.
 * - **Misc utilities** – deep-match, deep-merge, path navigation.
 */

// ---------------------------------------------------------------------------
// General utilities
// ---------------------------------------------------------------------------

/** Type guard: `value` is a non-null, non-array plain object. */
export const isRecord = (value: unknown): value is Record<string, unknown> => {
    return value !== null && typeof value === "object" && !Array.isArray(value);
};

/**
 * Walk `path` segments from `root`, creating intermediate objects (or using
 * the first element of an existing array) as needed.  Returns the leaf object.
 *
 * Used by extension setters to reach a nested target inside a resource.
 *
 * @example
 *   ensurePath(resource, ["contact", "telecom"])
 *   // → resource.contact.telecom (created if absent)
 */
export const ensurePath = (root: Record<string, unknown>, path: string[]): Record<string, unknown> => {
    let current: Record<string, unknown> = root;
    for (const segment of path) {
        if (Array.isArray(current[segment])) {
            const list = current[segment] as unknown[];
            if (list.length === 0) {
                list.push({});
            }
            current = list[0] as Record<string, unknown>;
        } else {
            if (!isRecord(current[segment])) {
                current[segment] = {};
            }
            current = current[segment] as Record<string, unknown>;
        }
    }
    return current;
};

// ---------------------------------------------------------------------------
// Deep match / merge
// ---------------------------------------------------------------------------

/**
 * Deep-merge `match` into `target`, mutating `target` in place.
 * Skips prototype-pollution keys.  Used internally by {@link applySliceMatch}.
 */
export const mergeMatch = (target: Record<string, unknown>, match: Record<string, unknown>): void => {
    for (const [key, matchValue] of Object.entries(match)) {
        if (key === "__proto__" || key === "constructor" || key === "prototype") {
            continue;
        }
        if (isRecord(matchValue)) {
            if (isRecord(target[key])) {
                mergeMatch(target[key] as Record<string, unknown>, matchValue);
            } else {
                target[key] = { ...matchValue };
            }
        } else {
            target[key] = matchValue;
        }
    }
};

/**
 * Shallow-clone `input` then deep-merge the slice discriminator values from
 * `match` on top, returning a complete slice element ready for insertion.
 *
 * @example
 *   applySliceMatch({ text: "hi" }, { coding: { code: "vital-signs", system: "…" } })
 *   // → { text: "hi", coding: { code: "vital-signs", system: "…" } }
 */
export const applySliceMatch = <T extends object>(input: Partial<T>, match: Partial<Record<keyof T, unknown>>): T => {
    const result = { ...input } as Record<string, unknown>;
    mergeMatch(result, match);
    return result as T;
};

/**
 * Recursively test whether `value` structurally contains everything in
 * `match`.  Arrays are matched with "every match item has a corresponding
 * value item" semantics; objects are matched key-by-key; primitives use `===`.
 *
 * This is the core discriminator check used to identify which array element
 * belongs to a given FHIR slice.
 */
export const matchesValue = (value: unknown, match: unknown): boolean => {
    if (Array.isArray(match)) {
        if (!Array.isArray(value)) {
            return false;
        }
        return match.every((matchItem) => value.some((item) => matchesValue(item, matchItem)));
    }
    if (isRecord(match)) {
        if (!isRecord(value)) {
            return false;
        }
        for (const [key, matchValue] of Object.entries(match)) {
            if (!matchesValue((value as Record<string, unknown>)[key], matchValue)) {
                return false;
            }
        }
        return true;
    }
    return value === match;
};

/**
 * Type guard that discriminates a raw extension input (with an `extension`
 * array) from a flat-API input object.  Using a custom type guard instead of
 * a bare `"extension" in args` lets TypeScript narrow *both* branches of the
 * union — the plain `in` check cannot eliminate a type whose `extension`
 * property is optional.
 */
export const isRawExtensionInput = <TRaw extends object>(input: object): input is TRaw => "extension" in input;

/**
 * Type guard that tests whether an unknown setter input is a raw Extension
 * (i.e. an object with a `url` property).  When `url` is provided, also
 * checks that the extension's URL matches the expected value.
 */
export const isExtension = <E extends { url: string }>(input: unknown, url?: string): input is E =>
    typeof input === "object" && input !== null && "url" in input && (url === undefined || input.url === url);

/**
 * Read a single typed value field from an Extension, returning `undefined`
 * when the extension itself is absent or the field is not set.
 *
 * This avoids the double-cast `(ext as Record<…>)?.field as T` that would
 * otherwise be needed for value fields not declared on the base Extension type.
 */
export const getExtensionValue = <T>(ext: { url?: string } | undefined, field: string): T | undefined => {
    if (!ext) return undefined;
    return (ext as Record<string, unknown>)[field] as T | undefined;
};

/**
 * Push an extension onto `target.extension`, creating the array if absent.
 */
export const pushExtension = <E extends { url?: string }>(target: { extension?: E[] }, ext: E): void => {
    (target.extension ??= []).push(ext);
};

// ---------------------------------------------------------------------------
// Extension helpers
// ---------------------------------------------------------------------------

/**
 * Read a complex (nested) FHIR extension into a plain key/value object.
 *
 * Each entry in `config` describes one sub-extension by URL, the name of its
 * value field (e.g. `"valueString"`), and whether it may repeat.
 *
 * @returns A record keyed by sub-extension URL, or `undefined` if the
 *          extension has no nested children.
 */
export const extractComplexExtension = <T = Record<string, unknown>>(
    extension: { extension?: Array<{ url?: string }> } | undefined,
    config: Array<{ name: string; valueField: string; isArray: boolean }>,
): T | undefined => {
    if (!extension?.extension) return undefined;
    const result: Record<string, unknown> = {};
    for (const { name, valueField, isArray } of config) {
        const subExts = extension.extension.filter((e) => e.url === name);
        if (isArray) {
            result[name] = subExts.map((e) => (e as Record<string, unknown>)[valueField]);
        } else if (subExts[0]) {
            result[name] = (subExts[0] as Record<string, unknown>)[valueField];
        }
    }
    return result as T;
};

// ---------------------------------------------------------------------------
// Slice helpers
// ---------------------------------------------------------------------------

/**
 * Remove discriminator keys from a slice element, returning only the
 * user-supplied portion.  Used by slice getters so callers see a clean object
 * without the fixed discriminator values baked in.
 */
export const stripMatchKeys = <T>(slice: object, matchKeys: string[]): T => {
    const result = { ...slice } as Record<string, unknown>;
    for (const key of matchKeys) {
        delete result[key];
    }
    return result as T;
};

/**
 * Wrap a flat input object under a choice-type key before inserting into a
 * slice.  For example, a Quantity value destined for `valueQuantity` is
 * wrapped as `{ valueQuantity: { ...input } }`.
 *
 * No-op when `input` is empty (the slice will contain only discriminator
 * defaults).
 */
export const wrapSliceChoice = <T>(input: object, choiceVariant: string): Partial<T> => {
    if (Object.keys(input).length === 0) return input as Partial<T>;
    return { [choiceVariant]: input } as Partial<T>;
};

/**
 * Inverse of {@link wrapSliceChoice}: strip discriminator keys, then hoist
 * the value inside `choiceVariant` up to the top level.
 *
 * @example
 *   unwrapSliceChoice(raw, ["code"], "valueQuantity")
 *   // removes "code", moves raw.valueQuantity.* to top level
 */
export const unwrapSliceChoice = <T>(slice: object, matchKeys: string[], choiceVariant: string): T => {
    const result = { ...slice } as Record<string, unknown>;
    for (const key of matchKeys) {
        delete result[key];
    }
    const variantValue = result[choiceVariant];
    delete result[choiceVariant];
    if (isRecord(variantValue)) {
        Object.assign(result, variantValue);
    }
    return result as T;
};

/**
 * Ensure that every required slice has at least a stub element in the array.
 * Each `match` is a discriminator pattern; if no existing item satisfies it,
 * a deep clone of the pattern is appended.
 *
 * Called in `createResource` so that required slices are always present even
 * when the caller omits them.
 */
export const ensureSliceDefaults = <T>(items: T[], ...matches: Record<string, unknown>[]): T[] => {
    for (const match of matches) {
        if (!items.some((item) => matchesValue(item, match))) {
            items.push(structuredClone(match) as T);
        }
    }
    return items;
};

/**
 * Cast an object literal to a FHIR resource type.  This centralises the single
 * `as unknown as T` that is unavoidable when constructing a resource from a
 * plain object (deep inheritance prevents direct structural compatibility).
 */
export const buildResource = <T>(obj: object): T => obj as unknown as T;

/**
 * Add `canonicalUrl` to `resource.meta.profile` if not already present.
 * Creates `meta` and `profile` when missing.
 */
export const ensureProfile = (resource: { meta?: { profile?: string[] } }, canonicalUrl: string): void => {
    const meta = (resource.meta ??= {});
    const profiles = (meta.profile ??= []);
    if (!profiles.includes(canonicalUrl)) profiles.push(canonicalUrl);
};

/**
 * Find or insert a slice element in `list`.  If an element matching `match`
 * already exists it is replaced in place; otherwise `value` is appended.
 */
export const setArraySlice = <T>(list: T[], match: Record<string, unknown>, value: T): void => {
    const index = list.findIndex((item) => matchesValue(item, match));
    if (index === -1) {
        list.push(value);
    } else {
        list[index] = value;
    }
};

/** Return the first element in `list` that satisfies the slice discriminator `match`. */
export const getArraySlice = <T>(list: readonly T[] | undefined, match: Record<string, unknown>): T | undefined => {
    if (!list) return undefined;
    return list.find((item) => matchesValue(item, match));
};

// ---------------------------------------------------------------------------
// Validation helpers
//
// Each function returns an array of human-readable error strings (empty = ok).
// Profile classes spread them all into a single array from `validate()`.
// ---------------------------------------------------------------------------

/** Checks that `field` is present (not `undefined` or `null`). */
export const validateRequired = (res: object, profileName: string, field: string): string[] => {
    const rec = res as Record<string, unknown>;
    return rec[field] === undefined || rec[field] === null
        ? [`${profileName}: required field '${field}' is missing`]
        : [];
};

/** Checks that a must-support field is populated (warning, not error). */
export const validateMustSupport = (res: object, profileName: string, field: string): string[] => {
    const rec = res as Record<string, unknown>;
    return rec[field] === undefined || rec[field] === null
        ? [`${profileName}: must-support field '${field}' is not populated`]
        : [];
};

/** Checks that `field` is absent (profiles may exclude base fields). */
export const validateExcluded = (res: object, profileName: string, field: string): string[] => {
    return (res as Record<string, unknown>)[field] !== undefined
        ? [`${profileName}: field '${field}' must not be present`]
        : [];
};

/** Checks that `field` structurally contains the expected fixed value. */
export const validateFixedValue = (res: object, profileName: string, field: string, expected: unknown): string[] => {
    return matchesValue((res as Record<string, unknown>)[field], expected)
        ? []
        : [`${profileName}: field '${field}' does not match expected fixed value`];
};

/**
 * Checks that the number of array elements matching `match` (a slice
 * discriminator) falls within [`min`, `max`].  Pass `max = 0` for unbounded.
 */
export const validateSliceCardinality = (
    res: object,
    profileName: string,
    field: string,
    match: Record<string, unknown>,
    sliceName: string,
    min: number,
    max: number,
): string[] => {
    const items = (res as Record<string, unknown>)[field] as unknown[] | undefined;
    const count = (items ?? []).filter((item) => matchesValue(item, match)).length;
    const errors: string[] = [];
    if (count < min) {
        errors.push(`${profileName}.${field}: slice '${sliceName}' requires at least ${min} item(s), found ${count}`);
    }
    if (max > 0 && count > max) {
        errors.push(`${profileName}.${field}: slice '${sliceName}' allows at most ${max} item(s), found ${count}`);
    }
    return errors;
};

/**
 * Checks that at least one of the listed choice-type variants is present.
 * E.g. `["effectiveDateTime", "effectivePeriod"]`.
 */
export const validateChoiceRequired = (res: object, profileName: string, choices: string[]): string[] => {
    const rec = res as Record<string, unknown>;
    return choices.some((c) => rec[c] !== undefined)
        ? []
        : [`${profileName}: at least one of ${choices.join(", ")} is required`];
};

/**
 * Checks that the value of `field` has a code within `allowed`.
 * Handles plain strings, Coding objects, and CodeableConcept objects.
 * Skips validation when the field is absent.
 */
export const validateEnum = (res: object, profileName: string, field: string, allowed: string[]): string[] => {
    const value = (res as Record<string, unknown>)[field];
    if (value === undefined || value === null) return [];
    if (typeof value === "string") {
        return allowed.includes(value)
            ? []
            : [`${profileName}: field '${field}' value '${value}' is not in allowed values`];
    }
    const rec = value as Record<string, unknown>;
    // Coding
    if (typeof rec.code === "string" && rec.system !== undefined) {
        return allowed.includes(rec.code)
            ? []
            : [`${profileName}: field '${field}' code '${rec.code}' is not in allowed values`];
    }
    // CodeableConcept
    if (Array.isArray(rec.coding)) {
        const codes = (rec.coding as Record<string, unknown>[]).map((c) => c.code as string).filter(Boolean);
        const hasValid = codes.some((c) => allowed.includes(c));
        return hasValid ? [] : [`${profileName}: field '${field}' has no coding with an allowed code`];
    }
    return [];
};

/**
 * Checks that a Reference field points to one of the `allowed` resource
 * types.  Extracts the type from the `reference` string (the part before
 * the first `/`).  Skips validation when the field or reference is absent.
 */
export const validateReference = (res: object, profileName: string, field: string, allowed: string[]): string[] => {
    const value = (res as Record<string, unknown>)[field];
    if (value === undefined || value === null) return [];
    const ref = (value as Record<string, unknown>).reference as string | undefined;
    if (!ref) return [];
    const slashIdx = ref.indexOf("/");
    if (slashIdx === -1) return [];
    const refType = ref.slice(0, slashIdx);
    return allowed.includes(refType)
        ? []
        : [`${profileName}: field '${field}' references '${refType}' but only ${allowed.join(", ")} are allowed`];
};

File: developer-experience/atomic-ehr-codegen-python-us-core-profiles/fhir_types/profile_helpers.py
"""
Runtime helpers for generated FHIR profile classes.

This file is copied verbatim into every generated Python output and imported by
profile modules. It provides:

- **Slice helpers** – match, get, set, and default-fill array slices defined by
  a FHIR StructureDefinition.
- **Extension helpers** – read complex (nested) FHIR extensions into plain dicts.
- **Choice-type helpers** – wrap/unwrap polymorphic ``value[x]`` fields so
  profile classes can expose a flat API.
- **Validation helpers** – lightweight structural checks that profile classes
  call from their ``validate()`` method.
- **Misc utilities** – deep-match, deep-merge, path navigation.

The helpers operate on plain ``dict`` / ``list`` structures. Profile classes
own a Pydantic resource instance (``self._resource``); when a helper needs the
underlying data, the profile passes ``self._resource.model_dump(by_alias=True,
exclude_none=True)`` or accesses model fields directly. All ``validate_*``
functions return ``list[str]`` so a profile's ``validate()`` can concatenate
them into a single errors / warnings list.
"""

from __future__ import annotations

import copy
from typing import Any, Iterable, Mapping, MutableMapping, MutableSequence, Sequence, TypeVar

from typing_extensions import TypeGuard

T = TypeVar("T")

# ---------------------------------------------------------------------------
# General utilities
# ---------------------------------------------------------------------------


def is_record(value: Any) -> TypeGuard[MutableMapping[str, Any]]:
    """True when ``value`` is a non-None mapping (dict-like, not a list)."""
    return isinstance(value, MutableMapping)


def ensure_path(root: MutableMapping[str, Any], path: Sequence[str]) -> MutableMapping[str, Any]:
    """Walk ``path`` from ``root``, creating intermediate dicts (or using the
    first element of an existing list) as needed. Returns the leaf mapping.

    Used by extension setters to reach a nested target inside a resource dict.
    """
    current: MutableMapping[str, Any] = root
    for segment in path:
        nxt = current.get(segment)
        if isinstance(nxt, list):
            if len(nxt) == 0:
                nxt.append({})
            current = nxt[0]
        else:
            if not isinstance(nxt, MutableMapping):
                nxt = {}
                current[segment] = nxt
            current = nxt
    return current


# ---------------------------------------------------------------------------
# Deep match / merge
# ---------------------------------------------------------------------------


def merge_match(target: MutableMapping[str, Any], match: Mapping[str, Any]) -> None:
    """Deep-merge ``match`` into ``target``, mutating ``target`` in place."""
    for key, match_value in match.items():
        if is_record(match_value):
            existing = target.get(key)
            if is_record(existing):
                merge_match(existing, match_value)
            else:
                target[key] = dict(match_value)
        else:
            target[key] = match_value


def apply_slice_match(input_obj: Mapping[str, Any], match: Mapping[str, Any]) -> dict[str, Any]:
    """Shallow-clone ``input_obj`` then deep-merge ``match`` on top, returning
    a complete slice element ready for insertion."""
    result: dict[str, Any] = dict(input_obj)
    merge_match(result, match)
    return result


def _get_key(obj: Any, key: str) -> Any:
    """Retrieve ``key`` from a dict-like or Pydantic-model-like object."""
    if is_record(obj):
        return obj.get(key)
    return getattr(obj, key, None)


def _model_get(value: Any, key: str) -> Any:
    """Get an attribute from a Pydantic model by Python name or field alias.

    Pydantic stores fields as snake_case attributes but slice match dicts use
    camelCase aliases (e.g. ``resourceType``). Try the direct attribute first;
    fall back to scanning ``model_fields`` for a matching alias.
    """
    direct = getattr(value, key, None)
    if direct is not None:
        return direct
    model_fields = getattr(type(value), "model_fields", None)
    if model_fields:
        for field_name, field_info in model_fields.items():
            if field_info.alias == key or field_info.serialization_alias == key:
                return getattr(value, field_name, None)
    return None


def matches_value(value: Any, match: Any) -> bool:
    """Recursively test whether ``value`` structurally contains everything in
    ``match``. Lists use "every match item has a corresponding value item"
    semantics; mappings are matched key-by-key; primitives use ``==``.

    Works with both plain dicts and Pydantic model instances.  When ``match``
    is a record and ``value`` is a list, returns ``True`` if any element in
    ``value`` satisfies the record match (handles nested array fields in FHIR
    discriminator patterns).

    Core discriminator check used to identify which array element belongs to a
    given FHIR slice.
    """
    if isinstance(match, list):
        if not isinstance(value, list):
            return False
        return all(any(matches_value(item, m_item) for item in value) for m_item in match)
    if is_record(match):
        if value is None:
            return False
        # Record match against a list: check any element matches
        if isinstance(value, list):
            return any(matches_value(item, match) for item in value)
        # Plain dict
        if is_record(value):
            for key, m_val in match.items():
                if not matches_value(value.get(key), m_val):
                    return False
            return True
        # Pydantic model (or any object with attributes) — use alias-aware lookup
        if hasattr(value, "__dict__"):
            for key, m_val in match.items():
                if not matches_value(_model_get(value, key), m_val):
                    return False
            return True
        return False
    return bool(value == match)


def is_extension(value: Any, url: str | None = None) -> bool:
    """True when ``value`` looks like a raw FHIR Extension (has a ``url``).
    When ``url`` is given, also checks the URL matches.
    Works with both plain dicts and Pydantic model instances."""
    ext_url = _get_key(value, "url") if (is_record(value) or hasattr(value, "__dict__")) else None
    if ext_url is None:
        return False
    return url is None or bool(ext_url == url)


def get_extension_value(ext: Any | None, field: str) -> Any:
    """Read a single typed value field from an Extension dict or Pydantic model,
    returning ``None`` when the extension itself is absent or the field is not set."""
    if ext is None:
        return None
    return _get_key(ext, field)


def push_extension(target: Any, ext: Any) -> None:
    """Push an extension onto ``target.extension`` (Pydantic model) or
    ``target['extension']`` (dict), creating the list if absent. ``ext`` may
    be either a dict-like mapping or a Pydantic model instance — mappings are
    shallow-copied, Pydantic models are stored as-is so attribute access and
    nested model instances are preserved."""
    lst = getattr(target, "extension", None) if hasattr(target, "__dict__") else target.get("extension")
    if not isinstance(lst, list):
        lst = []
    if hasattr(ext, "model_dump"):
        lst.append(ext)
    else:
        lst.append(dict(ext))
    if hasattr(target, "__dict__"):
        setattr(target, "extension", lst)
    else:
        target["extension"] = lst


# ---------------------------------------------------------------------------
# Extension helpers
# ---------------------------------------------------------------------------


def extract_complex_extension(
    extension: Any | None,
    config: Sequence[Mapping[str, Any]],
) -> dict[str, Any] | None:
    """Read a complex (nested) FHIR extension into a plain key/value dict.

    Each entry in ``config`` describes one sub-extension by ``name`` (URL),
    ``valueField`` (e.g. ``"valueString"``), and ``isArray``.

    Works with both plain dicts and Pydantic model instances.
    """
    if extension is None:
        return None
    sub_exts = _get_key(extension, "extension")
    if not isinstance(sub_exts, list):
        return None
    result: dict[str, Any] = {}
    for entry in config:
        name = entry["name"]
        value_field = entry["valueField"]
        is_array = bool(entry["isArray"])
        matched = [e for e in sub_exts if _get_key(e, "url") == name]
        if is_array:
            result[name] = [_get_key(e, value_field) for e in matched]
        elif matched:
            result[name] = _get_key(matched[0], value_field)
    return result


# ---------------------------------------------------------------------------
# Slice helpers
# ---------------------------------------------------------------------------


def strip_match_keys(slice_obj: Mapping[str, Any], match_keys: Sequence[str]) -> dict[str, Any]:
    """Remove discriminator keys from a slice element, returning only the
    user-supplied portion."""
    result = dict(slice_obj)
    for key in match_keys:
        result.pop(key, None)
    return result


def wrap_slice_choice(input_obj: Mapping[str, Any], choice_variant: str) -> dict[str, Any]:
    """Wrap a flat input dict under a choice-type key. No-op when ``input_obj``
    is empty."""
    if len(input_obj) == 0:
        return dict(input_obj)
    return {choice_variant: dict(input_obj)}


def unwrap_slice_choice(
    slice_obj: Mapping[str, Any],
    match_keys: Sequence[str],
    choice_variant: str,
) -> dict[str, Any]:
    """Inverse of :func:`wrap_slice_choice`: strip discriminator keys, then
    hoist the value inside ``choice_variant`` up to the top level."""
    result = dict(slice_obj)
    for key in match_keys:
        result.pop(key, None)
    variant_value = result.pop(choice_variant, None)
    if is_record(variant_value):
        for k, v in variant_value.items():
            result[k] = v
    return result


def ensure_slice_defaults(items: MutableSequence[Any], *matches: Mapping[str, Any]) -> MutableSequence[Any]:
    """Ensure that every required slice has at least a stub element in the
    array. If no existing item satisfies a ``match``, a deep clone of the
    pattern is appended."""
    for match in matches:
        if not any(matches_value(item, match) for item in items):
            items.append(copy.deepcopy(dict(match)))
    return items


def build_resource(resource_cls: type[T], /, **fields: Any) -> T:
    """Instantiate a Pydantic resource class from kwargs, dropping ``None``
    values so optional fields don't appear in the dump.

    Centralises construction so generators don't need to import every model.
    """
    cleaned = {k: v for k, v in fields.items() if v is not None}
    return resource_cls(**cleaned)


def ensure_profile(resource: Any, canonical_url: str) -> None:
    """Add ``canonical_url`` to ``resource.meta.profile`` if not already
    present. Works on both Pydantic models and plain dicts; creates ``meta``
    and ``profile`` when missing."""
    if isinstance(resource, MutableMapping):
        meta = resource.get("meta")
        if not isinstance(meta, MutableMapping):
            meta = {}
            resource["meta"] = meta
        profiles = meta.get("profile")
        if not isinstance(profiles, list):
            profiles = []
            meta["profile"] = profiles
        if canonical_url not in profiles:
            profiles.append(canonical_url)
        return
    # Pydantic model path
    meta = getattr(resource, "meta", None)
    if meta is None:
        # Try to construct a Meta from the model's annotation
        meta_field = type(resource).model_fields.get("meta") if hasattr(type(resource), "model_fields") else None
        if meta_field is not None and meta_field.annotation is not None:
            try:
                import types as _types
                import typing as _typing
                ann = meta_field.annotation
                # Unwrap Optional[T] / Union[T, None] / T | None to get the actual class
                origin = getattr(ann, "__origin__", None)
                if origin is _typing.Union or isinstance(ann, _types.UnionType):
                    args = [a for a in ann.__args__ if a is not type(None)]
                    if args:
                        ann = args[0]
                meta = ann(profile=[canonical_url])
                resource.meta = meta
                return
            except Exception:
                pass
        # Fallback: shouldn't happen for FHIR resources
        return
    profiles = getattr(meta, "profile", None)
    if profiles is None:
        meta.profile = [canonical_url]
    elif canonical_url not in profiles:
        profiles.append(canonical_url)


def set_array_slice(lst: MutableSequence[Any], match: Mapping[str, Any], value: Any) -> None:
    """Find or insert a slice element. If an element matching ``match``
    already exists it is replaced in place; otherwise ``value`` is appended."""
    for i, item in enumerate(lst):
        if matches_value(item, match):
            lst[i] = value
            return
    lst.append(value)


def get_array_slice(lst: Sequence[Any] | None, match: Mapping[str, Any]) -> Any:
    """Return the first element in ``lst`` that satisfies ``match``."""
    if lst is None:
        return None
    for item in lst:
        if matches_value(item, match):
            return item
    return None


def get_array_slices(lst: Sequence[Any] | None, match: Mapping[str, Any]) -> list[Any]:
    """Return all elements in ``lst`` that satisfy ``match``."""
    if lst is None:
        return []
    return [item for item in lst if matches_value(item, match)]


def set_array_slices(
    lst: MutableSequence[Any],
    match: Mapping[str, Any],
    values: Sequence[Any],
) -> None:
    """Remove all elements matching ``match``, then append ``values``."""
    indices = [i for i, item in enumerate(lst) if matches_value(item, match)]
    for i in reversed(indices):
        del lst[i]
    lst.extend(values)


# ---------------------------------------------------------------------------
# Validation helpers
#
# Each function returns a list of human-readable error strings (empty = ok).
# Profile classes concatenate them all into a single list inside validate().
# ---------------------------------------------------------------------------


def _get_field(res: Any, field: str) -> Any:
    if isinstance(res, Mapping):
        return res.get(field)
    return getattr(res, field, None)


def validate_required(res: Any, profile_name: str, field: str) -> list[str]:
    """Checks that ``field`` is present (not ``None``)."""
    return (
        [f"{profile_name}: required field '{field}' is missing"]
        if _get_field(res, field) is None
        else []
    )


def validate_must_support(res: Any, profile_name: str, field: str) -> list[str]:
    """Checks that a must-support field is populated (warning, not error)."""
    return (
        [f"{profile_name}: must-support field '{field}' is not populated"]
        if _get_field(res, field) is None
        else []
    )


def validate_excluded(res: Any, profile_name: str, field: str) -> list[str]:
    """Checks that ``field`` is absent."""
    return (
        [f"{profile_name}: field '{field}' must not be present"]
        if _get_field(res, field) is not None
        else []
    )


def validate_fixed_value(res: Any, profile_name: str, field: str, expected: Any) -> list[str]:
    """Checks that ``field`` structurally contains the expected fixed value."""
    actual = _get_field(res, field)
    return (
        []
        if matches_value(actual, expected)
        else [f"{profile_name}: field '{field}' does not match expected fixed value"]
    )


def validate_slice_cardinality(
    res: Any,
    profile_name: str,
    field: str,
    match: Mapping[str, Any],
    slice_name: str,
    min_count: int,
    max_count: int,
) -> list[str]:
    """Checks that the number of array elements matching ``match`` falls
    within ``[min_count, max_count]``. Pass ``max_count = 0`` for unbounded."""
    items = _get_field(res, field) or []
    if not isinstance(items, Iterable):
        items = []
    count = sum(1 for item in items if matches_value(item, match))
    errors: list[str] = []
    if count < min_count:
        errors.append(
            f"{profile_name}.{field}: slice '{slice_name}' requires at least {min_count} item(s), found {count}"
        )
    if max_count > 0 and count > max_count:
        errors.append(
            f"{profile_name}.{field}: slice '{slice_name}' allows at most {max_count} item(s), found {count}"
        )
    return errors


def validate_choice_required(res: Any, profile_name: str, choices: Sequence[str]) -> list[str]:
    """Checks that at least one of the listed choice-type variants is present."""
    if any(_get_field(res, c) is not None for c in choices):
        return []
    return [f"{profile_name}: at least one of {', '.join(choices)} is required"]


def validate_enum(res: Any, profile_name: str, field: str, allowed: Sequence[str]) -> list[str]:
    """Checks that the value of ``field`` has a code within ``allowed``.
    Handles plain strings, Coding, and CodeableConcept."""
    value = _get_field(res, field)
    if value is None:
        return []
    if isinstance(value, str):
        return (
            []
            if value in allowed
            else [f"{profile_name}: field '{field}' value '{value}' is not in allowed values"]
        )
    # Coding
    code = _get_field(value, "code")
    system = _get_field(value, "system")
    if isinstance(code, str) and system is not None:
        return (
            []
            if code in allowed
            else [f"{profile_name}: field '{field}' code '{code}' is not in allowed values"]
        )
    # CodeableConcept
    coding = _get_field(value, "coding")
    if isinstance(coding, list):
        codes = [_get_field(c, "code") for c in coding]
        codes = [c for c in codes if isinstance(c, str)]
        if any(c in allowed for c in codes):
            return []
        return [f"{profile_name}: field '{field}' has no coding with an allowed code"]
    return []


def validate_reference(res: Any, profile_name: str, field: str, allowed: Sequence[str]) -> list[str]:
    """Checks that a Reference field points to one of the ``allowed`` resource
    types. Extracts the type from the ``reference`` string (the part before
    the first ``/``)."""
    value = _get_field(res, field)
    if value is None:
        return []
    ref = _get_field(value, "reference")
    if not isinstance(ref, str):
        return []
    slash = ref.find("/")
    if slash == -1:
        return []
    ref_type = ref[:slash]
    if ref_type in allowed:
        return []
    return [
        f"{profile_name}: field '{field}' references '{ref_type}' but only {', '.join(allowed)} are allowed"
    ]
